In [1]:
import openml
import os
import pandas as pd
import math

## For loading data
from pathlib import Path
from typing import Union

In [2]:
def load_dataset(path: Union[Path, str]) -> pd.DataFrame:
    return pd.read_csv(path, index_col=0)


def load_rankings(path: Union[Path, str]) -> pd.DataFrame:
    out = pd.read_csv(path, index_col=0, header=[0, 1, 2, 3])
    out.columns.name = ("dataset", "model", "tuning", "scoring")
    return out

In [3]:
dir_data = '../../data/raw/'

# File names
filename_dataset = 'dataset.csv'

# Create paths for given files
filepath_dataset = os.path.join(dir_data, filename_dataset)

# Load data
dataset = load_dataset(filepath_dataset)

In [4]:
unique_datasets = dataset.dataset.unique()

# Collect data from datasets

In [5]:
%%time

list_count_target = []
list_count_features = []
list_min_number_of_categories = []
list_max_number_of_categories = []
list_sum_of_categories = []
list_count_cat_features = []
list_count_non_cat_feature = []
list_ratio = []
list_avg_categories_per_cat_feature = []
list_sum_of_target_categories = []
list_count_of_cat_targets = []
list_count_of_non_cat_targets = []

for dataset_id in unique_datasets:
    print(dataset_id)
    # Get dataset:
    dataset_3 = openml.datasets.get_dataset(dataset_id=int(dataset_id))
    
    # Count of target variables
    count_target = len(dataset_3.default_target_attribute.split(','))
    names_of_targets = dataset_3.default_target_attribute.split(',')
    
    # Count of all features
    count_features = len(dataset_3.features) - count_target
    
    # Count of nominal = categorical features
    # Min number of categories of categorical features
    # Max number of categories of categorical features
    # Sum of categories of all categorical features
    min_number_of_categories = math.inf
    max_number_of_categories = - math.inf
    sum_of_categories = 0
    sum_of_target_categories = 0
    count_of_cat_targets = 0
    count_of_non_cat_targets = 0
    type_and_count = {}
    
    for k in dataset_3.features:
        # Operations on features
        if dataset_3.features[k].name not in names_of_targets:
            # Update min and max number of categories per features
            if dataset_3.features[k].data_type == 'nominal':
                sum_of_categories += len(dataset_3.features[k].nominal_values)
                if len(dataset_3.features[k].nominal_values) < min_number_of_categories:
                    min_number_of_categories = len(dataset_3.features[k].nominal_values)
                if len(dataset_3.features[k].nominal_values) > max_number_of_categories:
                    max_number_of_categories = len(dataset_3.features[k].nominal_values)
            
            # Count features per data type
            dt = dataset_3.features[k].data_type
            if type_and_count.get(dt) is None:
                type_and_count[dt] = 1
            else:
                type_and_count[dt] = type_and_count[dt] + 1
        
        # Operations on targets
        else:
            if dataset_3.features[k].data_type == 'nominal':
                sum_of_target_categories += len(dataset_3.features[k].nominal_values)
                count_of_cat_targets += 1
            else:
                count_of_non_cat_targets += 1
        
    # Count of non-categorical features
    if type_and_count.get('nominal') == None:
        count_non_cat_feature = count_features
    else:
        count_non_cat_feature = count_features - type_and_count.get('nominal')
    
    # Ratio categorical / all features
    if type_and_count.get('nominal') == None:
        ratio = 0
    else:
        ratio = type_and_count.get('nominal') / count_features
    
    # Avg number of categories of categorical features
    if type_and_count.get('nominal') == None:
        avg_categories_per_cat_feature = 0
    else:
        avg_categories_per_cat_feature = sum_of_categories / type_and_count.get('nominal')
    
    
    # Update lists
    list_count_target.append(count_target)
    list_count_features.append(count_features)
    list_min_number_of_categories.append(min_number_of_categories)
    list_max_number_of_categories.append(max_number_of_categories)
    list_sum_of_categories.append(sum_of_categories)
    list_count_cat_features.append(type_and_count.get('nominal'))
    list_count_non_cat_feature.append(count_non_cat_feature)
    list_ratio.append(ratio)
    list_avg_categories_per_cat_feature.append(avg_categories_per_cat_feature)
    list_sum_of_target_categories.append(sum_of_target_categories)
    list_count_of_cat_targets.append(count_of_cat_targets)
    list_count_of_non_cat_targets.append(count_of_non_cat_targets)

# Create pandas df
data = {
    'dataset': unique_datasets,
    'count_target': list_count_target,
    'sum_of_target_categories': list_sum_of_target_categories,
    'count_of_cat_targets': list_count_of_cat_targets,
    'count_of_non_cat_targets': list_count_of_non_cat_targets,
    'count_features': list_count_features,
    'min_number_of_categories': list_min_number_of_categories,
    'max_number_of_categories': list_max_number_of_categories,
    'sum_of_categories': list_sum_of_categories,
    'count_cat_features': list_count_cat_features,
    'ratio_categorical_all': list_ratio,
    'avg_categories_per_cat_feature': list_avg_categories_per_cat_feature
}
meta_df = pd.DataFrame(data)

3
29
31
38
50
51
56
333
334
451
470
881
956
959
981
1037
1111
1112
1114
1169
1235
1461
1463
1486
1506
1511
1590
6332
23381
40536
40945
40981
40999
41005
41007
41162
41224


Could not download file from http://openml1.win.tue.nl/dataset41224/dataset_41224.pq: Bucket does not exist or is private.


42178
42343
42344
42738
42750
43098
43607
43890
43892
43896
43897
43900
43922
CPU times: user 103 ms, sys: 21.6 ms, total: 124 ms
Wall time: 276 ms


In [6]:
meta_df.head(50)

dataset  count_target  sum_of_target_categories  count_of_cat_targets   
0         3             1                         2                     1  \
1        29             1                         2                     1   
2        31             1                         2                     1   
3        38             1                         2                     1   
4        50             1                         2                     1   
5        51             1                         5                     1   
6        56             1                         2                     1   
7       333             1                         2                     1   
8       334             1                         2                     1   
9       451             1                         2                     1   
10      470             1                         2                     1   
11      881             1                         2                     1   
12      956             1                         2                     1   
13      959             1                         2                     1   
14      981             1                         2                     1   
15     1037             1                         2                     1   
16     1111             1                         2                     1   
17     1112             1                         2                     1   
18     1114             1                         2                     1   
19     1169             1                         2                     1   
20     1235             1                         2                     1   
21     1461             1                         2                     1   
22     1463             1                         2                     1   
23     1486             1                         2                     1   
24     1506             1                         2                     1   
25     1511             1                         2                     1   
26     1590             1                         2                     1   
27     6332             1                         2                     1   
28    23381             1                         2                     1   
29    40536             1                         2                     1   
30    40945             1                         2                     1   
31    40981             1                         2                     1   
32    40999             1                         3                     1   
33    41005             1                         3                     1   
34    41007             1                         3                     1   
35    41162             1                         2                     1   
36    41224             1                         2                     1   
37    42178             1                         0                     0   
38    42343             1                         2                     1   
39    42344             1                         2                     1   
40    42738             1                         0                     0   
41    42750             1                         0                     0   
42    43098             1                         0                     0   
43    43607             1                         0                     0   
44    43890             1                         2                     1   
45    43892             1                         2                     1   
46    43896             1                         2                     1   
47    43897             1                         2                     1   
48    43900             1                         2                     1   
49    43922             1                         2                     1   

    count_of_non_cat_targets  count_features  min_number_of_categories   
0 

# Analyze the new data 

Look for:
- constant columns $\rightarrow$ then drop them
- inf and -inf $\rightarrow$ replace them with 0
- Null values $\rightarrow$ depending on column
    - in 'count_cat_features' replace with 0 
    - so far I did not see any more NaN values

In [7]:
# Drop constant columns
clean_meta_df = meta_df.loc[:, (meta_df != meta_df.iloc[0]).any()] 

In [8]:
# Replace inf and -inf with -1
clean_meta_df = clean_meta_df.replace(to_replace={math.inf: 0, -math.inf: 0})

In [9]:
# Check for null values
clean_meta_df.isna().any()

dataset                           False
sum_of_target_categories          False
count_of_cat_targets              False
count_of_non_cat_targets          False
count_features                    False
min_number_of_categories          False
max_number_of_categories          False
sum_of_categories                 False
count_cat_features                 True
ratio_categorical_all             False
avg_categories_per_cat_feature    False
dtype: bool

In [10]:
clean_meta_df = clean_meta_df.fillna(0)

In [11]:
clean_meta_df.head(50)

dataset  sum_of_target_categories  count_of_cat_targets   
0         3                         2                     1  \
1        29                         2                     1   
2        31                         2                     1   
3        38                         2                     1   
4        50                         2                     1   
5        51                         5                     1   
6        56                         2                     1   
7       333                         2                     1   
8       334                         2                     1   
9       451                         2                     1   
10      470                         2                     1   
11      881                         2                     1   
12      956                         2                     1   
13      959                         2                     1   
14      981                         2                     1   
15     1037                         2                     1   
16     1111                         2                     1   
17     1112                         2                     1   
18     1114                         2                     1   
19     1169                         2                     1   
20     1235                         2                     1   
21     1461                         2                     1   
22     1463                         2                     1   
23     1486                         2                     1   
24     1506                         2                     1   
25     1511                         2                     1   
26     1590                         2                     1   
27     6332                         2                     1   
28    23381                         2                     1   
29    40536                         2                     1   
30    40945                         2                     1   
31    40981                         2                     1   
32    40999                         3                     1   
33    41005                         3                     1   
34    41007                         3                     1   
35    41162                         2                     1   
36    41224                         2                     1   
37    42178                         0                     0   
38    42343                         2                     1   
39    42344                         2                     1   
40    42738                         0                     0   
41    42750                         0                     0   
42    43098                         0                     0   
43    43607                         0                     0   
44    43890                         2                     1   
45    43892                         2                     1   
46    43896                         2                     1   
47    43897                         2                     1   
48    43900                         2                     1   
49    43922                         2                     1   

    count_of_non_cat_targets  count_features  min_number_of_categories   
0                          0              36                       2.0  \
1                          0              15                       2.0   
2                          0              20                       2.0   
3                          0              29                       1.0   
4                          0               9                       3.0   
5                          0              13                       2.0   
6                          0              16                       2.0   
7                          0               6                       2.0   
8                          0               6                       2.0   
9                          0               5      

In [12]:
clean_meta_df.describe()

dataset  sum_of_target_categories  count_of_cat_targets   
count     50.00000                 50.000000             50.000000  \
mean   18828.90000                  1.920000              0.900000   
std    20535.49872                  0.804071              0.303046   
min        3.00000                  0.000000              0.000000   
25%      956.75000                  2.000000              1.000000   
50%     1508.50000                  2.000000              1.000000   
75%    41939.50000                  2.000000              1.000000   
max    43922.00000                  5.000000              1.000000   

       count_of_non_cat_targets  count_features  min_number_of_categories   
count                 50.000000        50.00000                 50.000000  \
mean                   0.100000        46.74000                 13.540000   
std                    0.303046        82.95556                 72.671531   
min                    0.000000         5.00000                  0.000000   
25%                    0.000000         9.00000                  2.000000   
50%                    0.000000        16.00000                  2.000000   
75%                    0.000000        40.50000                  2.000000   
max                    1.000000       477.00000                513.000000   

       max_number_of_categories  sum_of_categories  count_cat_features   
count                 50.000000          50.000000           50.000000  \
mean                1947.620000        5682.360000           17.900000   
std                 5349.504783       17532.846483           23.628286   
min                    0.000000           0.000000            0.000000   
25%                    3.000000          21.000000            5.000000   
50%                   10.500000          55.000000            8.500000   
75%                  105.500000         301.250000           25.000000   
max                21838.000000       71504.000000          136.000000   

       ratio_categorical_all  avg_categories_per_cat_feature  
count              50.000000                       50.000000  
mean                0.547164                      259.324921  
std                 0.325347                      779.446982  
min                 0.000000                        0.000000  
25%                 0.255601                        2.640110  
50%                 0.563859                        4.403846  
75%                 0.828125                       12.281250  
max                 1.000000                     4376.600000

## Findings and manual checks

Manual checks: 
- Check dataset 1111, 1112, 1114. They have all the same values
  $\Rightarrow$ matches with the information given online (by visiting the website)
- Check dataset 40999, 41005, 41007. They have all the same values
  $\Rightarrow$ matches with the information given online (by visiting the website)

Findings: 
- Mostly we are dealing with classification problems. Only 5 datasets are designed for regression problems. 
- Only few datasets with > 100 features

## Save dataframe

In [13]:
clean_meta_df.to_csv('../../data/preprocessed/dataset_info.csv')